### PR TITLE
iso-image: normalize volumeID

### PR DESCRIPTION
**Pull request by worldofpeace**
###### Motivation for this change
The volumeID of nixos images will now be in the format of:
```
nixos-$EDITION-$RELEASE-$ARCH
nixos-minimal-20.09-x86-64-linux
```

We're trying to get nixos into osinfo-db, see https://gitlab.com/libosinfo/osinfo-db/-/merge_requests/107 and https://gitlab.com/libosinfo/osinfo-db/-/merge_requests/107#note_313094852 as motivation.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/installer/cd-dvd/installation-cd-graphical-plasma5-new-kernel.nix [hunk 1]
```diff
@@ -3,5 +3,7 @@
 {
   imports = [ ./installation-cd-graphical-plasma5.nix ];
 
+  isoImage.edition = "plasma5-new-kernel";
```
Comment (mmilata) [inline nixos/modules/installer/cd-dvd/installation-cd-graphical-plasma5-new-kernel.nix]: I think the whole volumeID is still limited to 32 characters, not sure what happens if we exceed this, either the ISO won't build or we lose (part of) the architecture.

**Review by mmilata — COMMENTED** [review]
Looking good:+1: 

Are we be able to distinguish unstable from release? If I understand correctly, current unstable has version == `20.09`.

Would be nice if someone (@wkennington?) checked whether #2838 is still a problem. Looking at the other ISOs in osinfo-db it probably shouldn't be.

File: nixos/modules/installer/cd-dvd/installation-cd-graphical-plasma5-new-kernel.nix (hunk 1 above)
Comment (worldofpeace) [reply nixos/modules/installer/cd-dvd/installation-cd-graphical-plasma5-new-kernel.nix]: Yeah, I don't think there's anything to get around it being limited to exactly 32 bytes https://wiki.osdev.org/ISO_9660#The_Primary_Volume_Descriptor. I wonder what exactly I can do.

**Comment by worldofpeace** [comment]
> Would be nice if someone (@wkennington?) checked whether #2838 is still a problem. Looking at the other ISOs in osinfo-db it probably shouldn't be.

I think the issue with windows will be fixed with using the joliet extension.

File: nixos/modules/installer/cd-dvd/installation-cd-graphical-plasma5-new-kernel.nix (hunk 1 above)
Comment (worldofpeace) [reply nixos/modules/installer/cd-dvd/installation-cd-graphical-plasma5-new-kernel.nix]: I did https://github.com/NixOS/nixpkgs/pull/83551/commits/69081e3dc742d2f7a617e216226c4c5eb36c51c0

**Comment by mmilata** [comment]
Installation in virt-manager working fine:)

**Comment by worldofpeace** [comment]
> Installation in virt-manager working fine:)

I made a vm in gnome-boxes. I'll see if I can install nixos to a usb and windows and start it eventually, since there was the issue in https://github.com/NixOS/nixpkgs/pull/2838.

**Comment by jtojnar** [comment]
Looking at the wiki page _Volume Identifier_ is `strD` type, which only supports `A-Z` `0-9` and `_` characters.

**Comment by worldofpeace** [comment]
> Looking at the wiki page _Volume Identifier_ is `strD` type, which only supports `A-Z`, `0-9` and `_` characters.

I guess a regex would work for this. Does it not support the dash?

**Comment by jtojnar** [comment]
No it does not: https://wiki.osdev.org/ISO_9660#String_format

**Comment by worldofpeace** [comment]
Perhaps https://wiki.osdev.org/ISO_9660#Rock_Ridge_and_Joliet has something to do with it. I see there's various improvements and extensions. For that matter, I see fedora has volume id's with dashes https://gitlab.com/libosinfo/osinfo-db/-/blob/master/data/os/fedoraproject.org/fedora-31.xml.in#L36. So perhaps there's something we're missing here.

**Comment by worldofpeace** [comment]
It seems ubuntu even has spaces https://gitlab.com/libosinfo/osinfo-db/-/blob/master/data/os/ubuntu.com/ubuntu-19.04.xml.in#L37 so maybe that documentation isn't an account of improvements

**Comment by jtojnar** [comment]
Briefly looking at the ECMA document, it looks like the Joliet format only adds enhanced volume descriptor, which mainly loosens the file and directory identifiers but not the volume identifier.

**Comment by mmilata** [comment]
Not sure if this part of the standard is mostly ignored or Joliet/Rockridge relaxes the constraint but seems minority of the ISOs found in osinfo-db conform:
```
$ ack -ch "<volume-id>.*</volume-id>"
2721
ack -ch "<volume-id>[A-Z0-9_]*</volume-id>"
188
```
(Ok, the content of the tag can be regex, but still only a minority are `strD`.)

Also from `man xorriso`:
>        -volid text
>              Specify the volume ID, which most operating systems will consider to be the volume name of the image or medium.
>              xorriso accepts any text up to 32 characters, but according to rarely obeyed specs stricter rules apply:
>              ECMA-119 demands ASCII characters out of [A-Z0-9_]. Like:
>                "IMAGE_23"
>              Joliet allows 16 UCS-2 characters. Like:
>                "Windows name"
>              Be aware that the volume id might get used automatically as the name of the mount point when the medium is inserted into a playful computer system.

**Comment by jtojnar** [comment]
Actually, I was reading the [ECMA doc](http://www.ecma-international.org/publications/files/ECMA-ST/Ecma-119.pdf) wrong. The _Volume Identifier_ described in 8.4.6 is only relevant for _Primary Volume Descriptor_. Joliet’s _Enhanced Volume Descriptor_ has its _Volume Identifier_ described in section 8.5.5, which specifies `d1-characters`:

> #### 7.4.2 c-characters
> The characters of the coded graphic character sets identified by the escape sequences in a Supplementary Volume Descriptor are referred to as c-characters.
> ##### 7.4.2.1 a1-characters
> A subset of the c-characters will be referred to as a1-characters. This subset shall be subject to agreement between the originator and the recipient of the volume.
> ##### 7.4.2.2 d1-characters
> A subset of the a1-characters will be referred to as d1-characters. This subset shall be subject to agreement between the originator and the recipient of the volume.

So if I am reading it right, we are indeed not limited by character range in Joliet format, only by the 32 byte size of the field.

**Comment by worldofpeace** [comment]
@jtojnar Great, thank you for the research. This kind of PR is one that could go wrong in the worst ways without.

**Pull request merged**